### PR TITLE
BlackoilModelEbos: improve nldd solver approach

### PR DESCRIPTION
**Pull request by akva2**
Separate out some more code into separate methods to improve higher-level readability.
Add enumeration of solver approach, and command line validation. In particular we get rid of an assert
on user provided data.

Sits on top of https://github.com/OPM/opm-simulators/pull/4738

### CODE REVIEW

**Comment by akva2** [comment]
jenkins build this please

**Comment by akva2** [comment]
jenkins build this please

File: opm/simulators/flow/BlackoilModelEbos.hpp [hunk 1]
```diff
@@ -1794,7 +1749,12 @@ namespace Opm {
             const auto& solution = ebosSimulator().model().solution(0);
 
             std::vector<int> domain_order(domains_.size());
-            if (param_.local_solve_approach_ == "gauss-seidel") {
+            switch (param_.local_solve_approach_) {
+            case DomainSolveApproach::Jacobi: {
+                std::iota(domain_order.begin(), domain_order.end(), 0);
+                break;
+            }
+            case DomainSolveApproach::GaussSeidel: {
```
Comment (bska) [inline opm/simulators/flow/BlackoilModelEbos.hpp:1757]: So this isn't **completely** equivalent to the original code.  That would return `[0, 1, ..., n-1]` for anything other than Gauss&ndash;Seidel.  This revised logic will return `[0, 1, ..., n-1]` only for the Jacobi method, and will return `[0, 0, ..., 0]` for anything else.  Maybe we should `throw` for unknown `local_solve_approach_` values?

**Review by bska — COMMENTED** [review]
Apart from a potential semantic change to the domain order function, this looks good to me.

File: opm/simulators/flow/BlackoilModelEbos.hpp (hunk 1 above)
Comment (akva2) [reply opm/simulators/flow/BlackoilModelEbos.hpp:1757]: it's an extensive switch.

Comment (bska) [reply opm/simulators/flow/BlackoilModelEbos.hpp:1757]: And by "anything else" I mean anything else that's **not** Gauss&ndash;Seidel.  Sorry for the confusion.

Comment (bska) [reply opm/simulators/flow/BlackoilModelEbos.hpp:1757]: > it's an extensive switch.

It's not.  That's one of the surprises of `enum class`.  The actual enumerator values are all of the values of the underlying integer types.

Comment (akva2) [reply opm/simulators/flow/BlackoilModelEbos.hpp:1757]: well okay, but it's still the only two values that can be assigned to the variable, we check and throw earlier in the process.

Comment (akva2) [reply opm/simulators/flow/BlackoilModelEbos.hpp]: added some default labels to retain original behavior (the other block had the opposite logic...)

**Comment by akva2** [comment]
jenkins build this please

**Pull request merged**